core/src/test/java/org/apache/iceberg/hadoop/TestCachingCatalog.java

```diff
@@ -252,9 +254,9 @@ public void testCacheExpirationEagerlyRemovesMetadataTables() throws IOException
     ticker.advance(HALF_OF_EXPIRATION);
     Assertions.assertThat(catalog.cache().asMap()).doesNotContainKey(tableIdent);
 
-    Arrays.stream(metadataTables(tableIdent)).forEach(metadataTable ->
-        Assert.assertFalse("When a data table expires, its metadata tables should expire regardless of age",
-            catalog.cache().asMap().containsKey(metadataTable)));
+    // Removal of metadata tables from cache is async, use awaitility
+    await().untilAsserted(() ->
+        Assertions.assertThat(catalog.cache().asMap()).doesNotContainKeys(metadataTables(tableIdent)));
```

Comment:
Question: Since you're introducing `Awaitility`, do you want to try to remove the call to `cleanUp` in the `TestableCachingCatalog` (the `cleanUp` call was added to handle the async expiration).
 https://github.com/apache/iceberg/blob/9c3e340cd67699c3d2499762d794ab1bc1ee7f45/core/src/test/java/org/apache/iceberg/TestableCachingCatalog.java#L46-L51
 
 This could likely be done in a separate PR though.

Reply:
Actually, this will balloon up the PR and this is an important fix. I'd say let's move everything in this test suite to awaitility in a separate PR (though the use of awaitlity is fine here).

That way we can review this change and get it out without having to review the much larger diff that will come from adding `awaitility` to a lot of this file.

Reply:
Thinking on it further, **Can we upgrade Caffeine**  to the next minor version (2.8.5) in this PR, which makes this `Awaitility` introduction unnecessary?

My reasoning is that the bug that causes you to have to use `Awaitility` here, is fixed byy upgrading to Caffeine 2.8.5. This indicates that the write path we're going through _also_ receives the fix and so we should provide that fix to our actual code too (even if we're not explicitly calling `cleanUp` ever, the bug could still affect us).

It's the[ only patch mentioned in the release notes](https://github.com/ben-manes/caffeine/releases/tag/v2.8.5) for 2.8.4 -> 2.8.5

We can and should still convert this to not call `cleanUp` in the `TestableCachingCache`, but I think it's best that the fix from 2.8.5 is also applied to the final code as it's a scheduling bug that could affect our new write path for this cache as well. This way also, we can introduce `Awaitility` in a follow up PR and focus only on this bug.

Reply:
The evictionListener replaces CacheWriter’s delete for an atomic action under the map’s computation. Since your previous code wrote back into the cache I think you would suffer the same problem if switching.

Since you don’t use a scheduler, I don’t see how that bug fix affects you.

You can use Caffeine.executor(Runnable::run) to disable async, which simplifies tests.

Reply:
Oh awesome thank you! You're right, after running the tests again several more times on 2.8.5, it didn't fix the issue. I just somehow got lucky and was able to get several successful runs in a row.

Appreciate the input regarding disabling async!

Reply:
We typically don't expect to see a _huge_ amount of cache usage. It's for caching tables that are accessed, and a job usually accesses a table via the catalog one time at the start of the job when query planning happens.

There are cases where they can be accessed several times (as mentioned in the issue), but it's usually a pretty limited number of values and times to use it.

Would it be a terrible idea to use `Caffeine.executor(Runnable::run)` to disable async in the actual production code? Ideally the tables we're removing in the `RemovalListener` would expire synchronously when the main table they reference expires.

Thanks again for your input and the great library @ben-manes!

Reply:
It would be perfectly fine to disable async in production code (Redhat's Infinispan does that, for example). The async is primarily because we don't know the cost of running user-supplied callbacks like the removal listener, which might impact user-facing latencies. The cache's own logic is very fast and strictly uses amortized O(1) algorithms.

Reply:
Great. Our work in the callback is relatively negligible (building a list of maybe and disabling async would be the desired behavior. Really appreciate the input and knowledge sharing! 😊

Reply:
just updated the code to use `Runnable::run` :smile: